### PR TITLE
Fix erroneous code action/lens suggestions in Daml Studio

### PR DESCRIPTION
**Pull request by dylant-da**
daml changes: https://github.com/digital-asset/daml/pull/15731
ghcide changes: https://github.com/digital-asset/daml-ghcide/pull/23
ghc changes: https://github.com/digital-asset/ghc/pull/150

Our GHC pipeline sometimes communicates something to the LF conversion layer by synthesizing special names that the LF converter then detects. For example, a `let` statement inside of a template is synthesized as `_templateLet_<templateName>` - the LF conversion stage then picks this up and erases it, using its definition to synthesize some DAML-LF specific code.

Unfortunately, code lenses and code actions operate on source code before LFConversion, so Daml Studio will pick up this "hidden code" and suggest it in some cases, e.g. to fill type holes.

This PR makes some changes to exclude `_templateLet_` and a few other generated constructs from Daml Studio's mechanisms.

Todo:
- [ ] Amend conversion and detection for all underscore types (templateLet, choice, interface_instance, requires, method, view, possibly others)
- [ ] update `ghc-lib/template-desugaring.md`

### CODE REVIEW

**Comment by garyverhaegen-da** [comment]
I'm not qualified to review the code, but I can confirm this (75f3ef2c447dbd189c40bb0c53aaeb3aa9e13b37) fixes the issue I raised in #15256.

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs [hunk 1]
```diff
@@ -239,11 +239,11 @@ getFctDocs ctx@DocCtx{..} (DeclData decl docs) = do
...
-    guard (not $ "_choice_" `T.isPrefixOf` packRdrName name)
-    guard (not $ "_interface_instance_" `T.isPrefixOf` packRdrName name)
-    guard (not $ "_requires_" `T.isPrefixOf` packRdrName name)
-    guard (not $ "_method_" `T.isPrefixOf` packRdrName name)
-    guard (not $ "_view_" `T.isPrefixOf` packRdrName name)
+    guard (not $ "_choice$_" `T.isPrefixOf` packRdrName name)
+    guard (not $ "_interface_instance$_" `T.isPrefixOf` packRdrName name)
+    guard (not $ "_requires$_" `T.isPrefixOf` packRdrName name)
+    guard (not $ "_method$_" `T.isPrefixOf` packRdrName name)
+    guard (not $ "_view$_" `T.isPrefixOf` packRdrName name)
```
Comment (akrmn) [inline compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs:246]: This module already imports stuff from `ghc-lib{,-parser}`, so I think we could use something like `guard (not . isDamlGenerated . rdrNameOcc $ name)` here

**Review by akrmn — APPROVED** [review]
Thank you Dylan!

File: compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs (hunk 1 above)
Comment (dylant-da) [reply compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs]: fixed

**Pull request merged**